### ISSUE TITLE
Option to record and playback received code data

### ISSUE THREAD

**Issue by AESilky** [MorseKOB/PyKOB#104]
To aid in learning and debugging, it would help to have an option to record the incoming code stream and an option to source (read) a recorded stream and play & decode it as if it was a live feed.

For learning, that would help by recording a live session with other operators and then be able to 'play it back' (multiple times) to be able to listen to it to learn.

For debugging, this will help in testing/improving decoding and (in the future) AI features for auto-correct.

**Comment by leskerr** [comment]
I totally agree this would be a worthwhile enhancement, for all the reasons mentioned. The Capture & Replay directory in the PyKOB repo contains some quick and dirty utilities along these lines, but integrating something into the MKOB app would be much nicer.

**Comment by AESilky** [comment]
For each packet received I think the recorded data should include:
1. Timestamp
2. Wire
3. Station
4. Code data (the timing info received in that packet)

I would like to save it in JSON format, as there are a lot of utilities that work with JSON in different ways.
We can easily parse the JSON file to pull out the code to be able to 'play' while also having the information to be able to display the station as MKOB currently does, along with the date+time it was originally sent.

@leskerr is there more that you think should be recorded?

**Comment by leskerr** [comment]
I can't think of any other information that needs to be captured. I don't know anything about JSON, but I'll take your word for it that it's a good format to use.

Are you picturing that the record and playback functions would be embedded as options in the MKOB client, or would they be independent programs?

For me, the most important use would be to display the timing data in the received code sequences when the code reader fails to translate something properly. This requires correlating the code reader's text with individual code packets, letter by letter. Doing this manually is _really_ tedious. Anything that can make this job easier would be a real plus.

It would also be useful to playback the code through the code reader at full speed, without sounding out the Morse for each character. That way we could tweak the code reader and get quick feedback on how well it worked, without having to wait for the actual Morse to be sounded.

**Comment by AESilky** [comment]
I would like to add them as options in MKOB and also update the Capture-Replay to be able to create/read the same data file.
An option to 'translate/decode only' could be available in MKOB and Replay so it wouldn't need to wait for the sounder.

I think an option in Replay should be able to show the timing data and the character together.

**Comment by leskerr** [comment]
Outstanding!

As far as I'm concerned, the old Capture/Replay programs are obsolete and can be removed from the project.

**Comment by leskerr** [comment]
One more comment: it could be useful to capture code that's sent on the local key, not just packets received from the server. This would also allow for capturing transmissions that were sent years ago over dialup Morse and archived on cassette tapes (since digitized).

**Comment by pwdirks** [comment]
I'm not sure what advantages JSON offers here (the data is not exactly structured - do you mean to associate "Station" etc. as JSON keys?) but if you do use JSON, you could use the timestamp (which would need to be sufficiently precise to be unique, or artificially made greater than the preceding one by adding +1 time unit) as the key to every record.

Do you mean to capture / replay a single wire?  In that case you could make the top level a single wire and underneath there have {timestamp: {station, data}} records.  Even if the first attempt is single-wire we might want to plan for simultaneous recording of a number of wires, and playback of same?

**Comment by AESilky** [comment]
@leskerr I plan to change Capture/Replay to use the same classes to record and play back the data as I'll use in MKOB. That way I can play back, decode, and debug in a CLI app rather than having to debug in a GUI app.

I am planning to be able to record the local key.

@pwdirks As you suggest, I am using JSON to be able to create and save a 'record' for each data packet. It includes the wire, station, timestamp, local/remote flag, sequence number, and the timing data. The json module also includes methods that make it easy to write the data out and read it in. Having this json also makes it easier to read manually and pick out a section of it that you want to analyze. It is also something that can be put into a nosql db to be able to query - say for communication on a given wire, day, time.

**Comment by AESilky** [comment]
@leskerr What exactly is recorded on the tapes and digitized versions?

**Comment by leskerr** [comment]
> @leskerr What exactly is recorded on the tapes and digitized versions?

Bell 103 modem tones

**Comment by AESilky** [comment]
Thanks @leskerr I manually processed one using Audacity into something with dots-dashes and silence. It made it much easier to listen too. I'm working on the save/replay and when that is working I'll work on something that can translate the recordings into the 'save' format.

That should let us play them back just like code from the wire.

**Comment by leskerr** [comment]
> Thanks @leskerr I manually processed one using Audacity into something with dots-dashes and silence. It made it much easier to listen too. I'm working on the save/replay and when that is working I'll work on something that can translate the recordings into the 'save' format.
> 
> That should let us play them back just like code from the wire.

Wonderful!

**Comment by AESilky** [comment]
Here is a test output from the Recorder class. This is a local recording of "HI" "THIS IS A TEST":

% cat test.1597214919028.json 
{"ts": 1597214919028, "source": 1, "station": "Test Recorder", "wire": -1, "type": 1, "code": [-1000, 2, -1000, 60, -60, 60, -60, 60, -60, 60, -180, 60, -60, 60, -1000, 1]}
{"ts": 1597214919029, "source": 1, "station": "Test Recorder", "wire": -1, "type": 1, "code": [-460, 180]}
{"ts": 1597214919029, "source": 1, "station": "Test Recorder", "wire": -1, "type": 1, "code": [-230, 60, -60, 60, -60, 60, -60, 60]}
{"ts": 1597214919030, "source": 1, "station": "Test Recorder", "wire": -1, "type": 1, "code": [-230, 60, -60, 60]}
{"ts": 1597214919030, "source": 1, "station": "Test Recorder", "wire": -1, "type": 1, "code": [-230, 60, -60, 60, -60, 60]}
{"ts": 1597214919030, "source": 1, "station": "Test Recorder", "wire": -1, "type": 1, "code": []}
{"ts": 1597214919030, "source": 1, "station": "Test Recorder", "wire": -1, "type": 1, "code": [-460, 60, -60, 60]}
{"ts": 1597214919030, "source": 1, "station": "Test Recorder", "wire": -1, "type": 1, "code": [-230, 60, -60, 60, -60, 60]}
{"ts": 1597214919031, "source": 1, "station": "Test Recorder", "wire": -1, "type": 1, "code": []}
{"ts": 1597214919031, "source": 1, "station": "Test Recorder", "wire": -1, "type": 1, "code": [-460, 60, -60, 180]}
{"ts": 1597214919031, "source": 1, "station": "Test Recorder", "wire": -1, "type": 1, "code": []}
{"ts": 1597214919031, "source": 1, "station": "Test Recorder", "wire": -1, "type": 1, "code": [-460, 180]}
{"ts": 1597214919031, "source": 1, "station": "Test Recorder", "wire": -1, "type": 1, "code": [-230, 60]}
{"ts": 1597214919031, "source": 1, "station": "Test Recorder", "wire": -1, "type": 1, "code": [-230, 60, -60, 60, -60, 60]}
{"ts": 1597214919032, "source": 1, "station": "Test Recorder", "wire": -1, "type": 1, "code": [-230, 180]}

The information is recorded in packets in a JSON structure that includes:
1. Timestamp: "ts"
2. Source (`local`/`wire`): "source"
3. Station ID: "station"
4. Wire Number: "wire"
5. Code type: "type"
6. Code Sequence (key timing information): "code"

@leskerr I haven't done playback yet, but does the above JSON seem to record what you think will be needed? For recording the saved audio/modem recordings, I could add an option to provide an 'origin' offset time-stamp such that the 'ts' value would be from the date specified (the date the modem audio was recorded). Can you think of anything else that people would want to know.

**Comment by AESilky** [comment]
Here is a recording from 'Clock' run for a few minutes, including a 'text' transmission at a quarter after.

```
{"ts": 1597219439467, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-460, 180]}
{"ts": 1597219440120, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 60, -60, 60, -60, 60, -60, 60]}
{"ts": 1597219440409, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 60]}
{"ts": 1597219440409, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": []}
{"ts": 1597219441049, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-460, 180]}
{"ts": 1597219441460, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 60, -60, 60]}
{"ts": 1597219442109, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 180, -60, 180]}
{"ts": 1597219442399, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 60]}
{"ts": 1597219442399, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": []}
{"ts": 1597219443040, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-460, 60, -60, 60]}
{"ts": 1597219443570, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 60, -60, 60, -60, 60]}
{"ts": 1597219443570, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": []}
{"ts": 1597219444690, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-460, 60, -60, 180, -60, 180, -60, 60]}
{"ts": 1597219444690, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": []}
{"ts": 1597219445449, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-460, 60, -180, 60]}
{"ts": 1597219446340, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 60, -60, 60, -60, 60, -60, 180, -60, 60]}
{"ts": 1597219560526, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 180]}
{"ts": 1597219561176, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 60, -60, 60, -60, 60, -60, 60]}
{"ts": 1597219561467, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 60]}
{"ts": 1597219561467, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": []}
{"ts": 1597219562105, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-460, 180]}
{"ts": 1597219562516, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 60, -60, 60]}
{"ts": 1597219563166, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 180, -60, 180]}
{"ts": 1597219563456, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 60]}
{"ts": 1597219563457, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": []}
{"ts": 1597219564096, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-460, 60, -60, 60]}
{"ts": 1597219564625, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 60, -60, 60, -60, 60]}
{"ts": 1597219564625, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": []}
{"ts": 1597219565746, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-460, 60, -60, 180, -60, 180, -60, 60]}
{"ts": 1597219565746, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": []}
{"ts": 1597219566506, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-460, 60, -180, 60]}
{"ts": 1597219567397, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 60, -60, 60, -60, 60, -60, 60, -60, 60, -60, 60]}
{"ts": 1597219740576, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 180]}
{"ts": 1597219741226, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 60, -60, 60, -60, 60, -60, 60]}
{"ts": 1597219741517, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 60]}
{"ts": 1597219741517, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": []}
{"ts": 1597219742156, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-460, 180]}
{"ts": 1597219742566, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 60, -60, 60]}
{"ts": 1597219743215, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 180, -60, 180]}
{"ts": 1597219743505, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 60]}
{"ts": 1597219743505, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": []}
{"ts": 1597219744147, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-460, 60, -60, 60]}
{"ts": 1597219744675, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 60, -60, 60, -60, 60]}
{"ts": 1597219744675, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": []}
{"ts": 1597219745795, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-460, 60, -60, 180, -60, 180, -60, 60]}
{"ts": 1597219745795, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": []}
{"ts": 1597219746557, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-460, 60, -180, 60]}
{"ts": 1597219747446, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 180, -60, 60, -60, 60, -60, 180]}
{"ts": 1597219920629, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 180]}
{"ts": 1597219921280, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 60, -60, 60, -60, 60, -60, 60]}
{"ts": 1597219921570, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 60]}
{"ts": 1597219921570, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": []}
{"ts": 1597219922206, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-460, 180]}
{"ts": 1597219922617, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 60, -60, 60]}
{"ts": 1597219923269, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 180, -60, 180]}
{"ts": 1597219923560, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 60]}
{"ts": 1597219923560, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": []}
{"ts": 1597219924199, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-460, 60, -60, 60]}
{"ts": 1597219924729, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 60, -60, 60, -60, 60]}
{"ts": 1597219924729, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": []}
{"ts": 1597219925849, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-460, 60, -60, 180, -60, 180, -60, 60]}
{"ts": 1597219925849, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": []}
{"ts": 1597219926970, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-460, 60, -60, 180, -60, 180, -60, 60]}
{"ts": 1597219927859, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 60, -60, 60, -60, 180, -60, 60, -60, 60]}
{"ts": 1597220101043, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 180]}
{"ts": 1597220101692, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 60, -60, 60, -60, 60, -60, 60]}
{"ts": 1597220101983, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 60]}
{"ts": 1597220101983, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": []}
{"ts": 1597220102622, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-460, 180]}
{"ts": 1597220103032, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 60, -60, 60]}
{"ts": 1597220103683, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 180, -60, 180]}
{"ts": 1597220103972, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 60]}
{"ts": 1597220103972, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": []}
{"ts": 1597220104613, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-460, 60, -60, 60]}
{"ts": 1597220105142, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 60, -60, 60, -60, 60]}
{"ts": 1597220105142, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": []}
{"ts": 1597220105902, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-460, 60, -60, 180]}
{"ts": 1597220105903, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": []}
{"ts": 1597220106903, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-460, 60, -60, 60, -60, 180, -60, 60]}
{"ts": 1597220107552, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 60, -60, 60, -60, 180]}
{"ts": 1597220108082, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 60, -60, 180]}
{"ts": 1597220108733, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 60, -180, 60, -60, 60]}
{"ts": 1597220109140, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 180]}
{"ts": 1597220109433, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 60]}
{"ts": 1597220110082, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 60, -180, 60, -60, 60]}
{"ts": 1597220110082, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": []}
{"ts": 1597220111082, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-460, 60, -60, 60, -60, 60, -60, 60, -60, 60]}
{"ts": 1597220111612, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 60, -60, 180]}
{"ts": 1597220112142, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 60, -60, 60, -60, 60]}
{"ts": 1597220112553, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 180]}
{"ts": 1597220112554, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": []}
{"ts": 1597220113312, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-460, 60, -180, 60]}
{"ts": 1597220113842, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 180, -60, 60]}
{"ts": 1597220114132, "source": 1, "station": "Clock", "wire": -1, "type": 1, "code": [-230, 60]}
```

**Comment by leskerr** [comment]
This is wonderful! I can't think of any more information that needs to be captured. Here are a few comments:

- I know I'll want to correlate something I see in the code reader window (maybe a poor translation of the Morse, for example) with a particular point in the JSON file so I can analyze the timing.  It would be nice if there was some way to retrieve a time stamp from the reader window text, but I have no idea how to do this.
- Strictly speaking, the code sequences should be in parentheses, not brackets, since they're Python sequences, not lists. Maybe this is dictated by JSON syntax, which I'm not familiar with.
- I wonder where those empty code sequences are coming from. They're occurring at word breaks.

I'm looking forward to the playback feature. Capturing Greg's recordings seems challenging, but if anyone can do it, @AESilky, you can!

**Comment by leskerr** [comment]
> * Strictly speaking, the code sequences should be in parentheses, not brackets, since they're Python sequences, not lists. Maybe this is dictated by JSON syntax, which I'm not familiar with.

I meant to say 'Python _tuples_', not 'Python sequences'.

**Comment by AESilky** [comment]
I can create a custom encoder/decoder if you think it is important. This is the default encoding.

What if we wait until I implement playback and see if it makes a difference? If it does, or you would prefer 'parentheses' I can add a custom encoder/decoder.

**Comment by leskerr** [comment]
Brackets are just fine. No problem.

**Comment by AESilky** [comment]
For playback, I'm planning to add a parameter for `condense_dead_air` that will take a `seconds` value. Given a value of `5` it would remove any code gap greater than 5 seconds.

Used on the recording of 'Clock' it would remove the gaps between the code at each interval.

@leskerr @pwdirks What do you think? Should I have an option to replace it with a 'given' space time (maybe with a default of 5 seconds).

**Comment by AESilky** [comment]
Of course the default would be to play back in real-time. But in looking at the modem recordings, there are times when the line is 'dead' for many seconds/minutes, so removing those gaps seems like a valuable option.

**Comment by pwdirks** [comment]
I'm torn between the simplicity of this format (it's 1:1 with UDP packets, I guess?) vs. the inefficiency of all that repeated metadata.  Of the items you list to be recorded (which seems quite complete):

> Timestamp: "ts"
> Source (local/wire): "source"
> Station ID: "station"
> Wire Number: "wire"
> Code type: "type"
> Code Sequence (key timing information): "code"

I'm not sure what "Code type" you mean to capture besides "code" but you COULD, by nesting them in order of least-to-most likely to change, structure and compact the data a bit:

{ [ "source": "local"
    "capture": [
        "wire": 107,
        "activity": [
            { "station": "AB",
               "message": [
                    { "ts": 1597219439467, type: "code", "code": [-460, 180]},
                    { "ts": 1597219440120, type: "code", "code": [-230, 60, -60, 60, -60, 60, -60, 60]},
                    { "ts": 1597219440409, type: "code", "code": [-230, 60]},
                    { "ts": 1597219440409, type: "code", "code": []},
                    { "ts": 1597219441049, type: "code", "code": [-230, 60, -60, 60]}
                ]
            },
            { "station": "XY",
                "message": [
                    { "ts": 1597219441460, type: "code", "code": [-460, 180]},
                    { "ts": 1597219442109, type: "code", "code": [-230, 60, -60, 60]},
                    { "ts": 1597219442399, type: "code", "code": [-230, 60]}
                ]
            }
        ]
    ]
}

into a sequence of sources (usually 1), which capture a sequence of wires (usually 1), which have a stream of "activity" structures, one station for however long it transmits, represented as a series of timestamp/type/code tuples?

Of course whitespace could be suppressed by default.  In this day and age, space efficiency of the representation is not really an overriding concern, though, although perhaps readability is but perhaps re-ordering the left-to-right of the fields to match the least-to-most changing (except perhaps the timestamp) might also achieve the same effect of making it easy to pick out one message.

Anyway, I can't decide which way is best - heavily structured or strictly sequential.  Maybe, if nothing else, prefixing the capture with some information about the format ("capture-version" or "capture-format") might give us flexibility to change our mind in the future anyway?

**Comment by pwdirks** [comment]
Sigh - it looks like there's no way to comment that doesn't lose the formatting and makes the result completely illegible but try cutting-and-pasting the text into an editor and pretty-print it.  Let me see if I can include a screen-shot of how it looks to me.

![image](https://user-images.githubusercontent.com/48630827/90045846-7aaa5b00-dc84-11ea-92c0-bedb93d80c72.png)

**Comment by AESilky** [comment]
As you mention, storage isn't as great of driver as it used to be...
Given that, I wanted to make each line contain the information needed to be able to play back with good information in the display in the case that you cut a section out in the middle of a recording file.

I can nest information, but that would most likely break a playback if someone grabbed a few lines from the middle of a file.

If you don't think that will be common, I can easily adjust to have a nested structure.

**Comment by leskerr** [comment]
> For playback, I'm planning to add a parameter for `condense_dead_air` that will take a `seconds` value. Given a value of `5` it would remove any code gap greater than 5 seconds.
> 
> Used on the recording of 'Clock' it would remove the gaps between the code at each interval.
> 
> @leskerr @pwdirks What do you think? Should I have an option to replace it with a 'given' space time (maybe with a default of 5 seconds).

`condense_dead_air` would be a useful playback option. If you made the default to be 'on' with a gap of 5 seconds, I'd probably never have reason to change it.

**Comment by pwdirks** [comment]
+1.  Exactly what Les said.

If on playback it flashed a "..." or inserted that into the display buffer, that'd be a cool signal that "some time passed"?

**Comment by AESilky** [comment]
I can add '... [12s]' to indicate that 12 seconds passed, '... [1m5s] to indicate that a minute and 5 seconds passed.

**Comment by AESilky** [comment]
Of course, that would be a callback that other classes could hook into.
The Recorder class doesn't have any access to display information in the apps using it.

**Comment by pwdirks** [comment]
Exactly.  Just some way of indicating to the client that time passed (or, I suppose, depending on the playback interface, that might be obvious to the client from the timestamps anyway).

**Comment by AESilky** [comment]
So, a `Playback` app might be a CLI that plays back recordings, while MKOB might have an option to playback a recording as though it were live. In the 'Playback' case, it would be CLI and need to display to the console. In the MKOB case, it would want to display in the windows as if it were live.

**Comment by AESilky** [comment]
I think implementing a call back when the 'dead-air'was exceeded and a new code sequence was received would accomplish the goal.

**Comment by AESilky** [comment]
Then it is up to the calling app to display as appropriate.

**Comment by AESilky** [comment]
@pwdirks code type is American or International. Maybe that's not needed for a trained listener, but I think it is for MKOB to be able to decode it.

**Comment by AESilky** [comment]
I can reorder the fields. I'm still leaning toward having each line stand on its own, rather than nesting information. I don't think many people will be reading a recording file, so the ability to take a chunk from anywhere in the file and play it back seems a higher priority to me.

**Comment by AESilky** [comment]
@pwdirks - GH comments are markdown, so you can use a code block to keep your formatting...

```
{
    [
        "source": "local" "capture": [
            "wire": 107,
            "activity": [
                {
                    "station": "AB",
                    "message": [
                        {
                            "ts": 1597219439467, type: "code",
                            "code": [
                                -460,
                                180
                            ]
                        },
                        {
                            "ts": 1597219440120, type: "code",
                            "code": [
                                -230,
                                60,
                                -60,
                                60,
                                -60,
                                60,
                                -60,
                                60
                            ]
                        },
                        {
                            "ts": 1597219440409, type: "code",
                            "code": [
                                -230,
                                60
                            ]
                        },
                        {
                            "ts": 1597219440409, type: "code",
                            "code": []
                        },
                        {
                            "ts": 1597219441049, type: "code",
                            "code": [
                                -230,
                                60,
                                -60,
                                60
                            ]
                        }
                    ]
                },
                {
                    "station": "XY",
                    "message": [
                        {
                            "ts": 1597219441460, type: "code",
                            "code": [
                                -460,
                                180
                            ]
                        },
                        {
                            "ts": 1597219442109, type: "code",
                            "code": [
                                -230,
                                60,
                                -60,
                                60
                            ]
                        },
                        {
                            "ts": 1597219442399, type: "code",
                            "code": [
                                -230,
                                60
                            ]
                        }
                    ]
                }
            ]
        ]
    }
```

**Comment by AESilky** [comment]
For reordering the data, I like the timestamp first and the code last, since the timestamp is what orders the data and the code is of varying length. For the rest of the fields I can reorder to type (won't change for a single recording), wire, station, source.

```
{"ts": 1597220114132, "type": 1, "wire": -1, "station": "Clock", "source": 1, "code": [-230, 60]}
```
Of course, the order is only relevant for someone reading the file. When the code reads it, it doesn't matter. If file size seems important, I can shorten the identifiers to: 'ts', 't', 'w', 's', 'o' (origin), 'c'
That would cut out 20 characters per line.

**Comment by leskerr** [comment]
> I don't think many people will be reading a recording file, so the ability to take a chunk from anywhere in the file and play it back seems a higher priority to me.

Actually, I've been scanning those recording files by eye. I've had a lot of practice at reading code sequences like that, for various reasons (this time mostly out of curiosity). I don't have trouble with the current format, because the fields before the code sequence are all fixed width, so the code sequences line up nicely. I view the file in a window sized wide enough so the lines don't wrap.

**Comment by AESilky** [comment]
The station does vary in width (as seen in the Wire 11 recording from this evening), but that helps identify when the sender changed.

**Comment by leskerr** [comment]
Yes, that's actually a bonus. :-)

**Comment by AESilky** [comment]
In VSCode the file looks like this and the side bar (right side) makes it easy to see where the station changed.

![B7DA7D9A-0C72-4223-BAE4-23A5936895A8_1_105_c](https://user-images.githubusercontent.com/19332027/90091996-9fcab800-dcdc-11ea-8a09-8820d855f563.jpeg)

**Comment by AESilky** [comment]
I do think I'll reduce the identifiers as mentioned. That will help reduce the width in the editor to get more of the code in without panning.

**Comment by AESilky** [comment]
This is what that looks like:

![B376F43A-7EF0-466D-9FE5-A4A9058DDC4F_1_105_c](https://user-images.githubusercontent.com/19332027/90092552-10260900-dcde-11ea-8bff-1454ce590885.jpeg)

**Comment by leskerr** [comment]
And this is what the capture file looks like in Notepad:

![Capture](https://user-images.githubusercontent.com/63867213/90093155-b8889d00-dcdf-11ea-8681-07af2d35959a.png)

Not as colorful, but even the long lines fit with room to spare.

**Comment by AESilky** [comment]
Very good. I will still try the reduced identifiers, as that will make the file smaller as well, and I think it will still be very readable.

**Comment by pwdirks** [comment]
I'd drop the "code" attribute altogether - that's really for the listener to know or figure out.  I'd hate to record something "in the wrong format" and then not be able to replay it by fixing the error.  On a real wire all I hear is click/clack - no indication what the semantics are.  And if I can fix the code type later, why bother specifying it in the first place?
I'd rather see MKOB (a) default in some intelligent way (maybe to American Morse except for certain known wires?) and (b) offer a way to override the default to get what you need.  Then record w/o indication of "code", but maybe apply the same default on playback based on "wire" treatment?
I guess generally I'd expect playback/record to capture the "raw" information on the wire(s), not the interpretation.

**Comment by AESilky** [comment]
Great input. I will drop that.

**Comment by AESilky** [comment]
I have pushed a version of playback. It requires the updated format, so I've attached a wire-11 file with that. It supports the '-h/--help' argument.

Try it with: python3 Capture-Replay/play.py --datetime MKOB/MKOB.1597281216622.edit.json
(adjusting for your mkob file path)

[MKOB.1597281216622.edit.json.zip](https://github.com/MorseKOB/PyKOB/files/5068308/MKOB.1597281216622.edit.json.zip)

**Comment by AESilky** [comment]
`Play.py`

**Comment by AESilky** [comment]
Branch: es-104-record-playback

**Comment by AESilky** [comment]
@leskerr I commented out the sleep and tested with the Clock recording where it announced every 3 minutes. Without the sleep it played the whole file without the 3 minute gaps.

So, I think the option for a sleep is needed in order to be able to play a recording as it was in real time (though I agree that you would rarely want to do that). I will look at the playback in more detail and try to come up with a better algorithm for incorporating the sleep.

**Comment by AESilky** [comment]
Adding the naive sleep back in made the Clock recording playback with the 3 minute pauses - so I think I need to so something to allow a correct playback in the case that you don't want to condense deadtime.

However, I did notice that a 'after pause' line was printed before the delay - so that could be a clus to a needed change. Here is an example with the current code:
```
Thu Aug 13 03:27:07 2020:  {"ts": 1597314427699, "w": -1, "s": "Clock", "o": 1, "c": [-230, 180, -60, 180, -60, 60, -60, 60]}
Thu Aug 13 03:30:00 2020:  {"ts": 1597314600885, "w": -1, "s": "Clock", "o": 1, "c": [-230, 180]}
Sleep:  173.186
Thu Aug 13 03:30:01 2020:  {"ts": 1597314601545, "w": -1, "s": "Clock", "o": 1, "c": [-230, 60, -60, 60, -60, 60, -60, 60]}
```
I wouldn't expect to see the '03:30:00' line before the sleep. I will look at that first.

**Comment by leskerr** [comment]
I think if you do a sleep only for gaps greater than one second, that'll work well. `KOB.sounder` will handle the pacing accurately for short gaps, where the timing is critical, and `sleep` will keep things properly synchronized when there are longer gaps.

**Comment by AESilky** [comment]
I will look at the play code and see what amount of time should trigger the sleep (if enabled). Of course we want the option to condense dead=time, so I'll make sure to incorporate that.

I'm thinking of an option that lets you condense dead-time > x to y (x and y being in seconds. With defaults of x=5 and y=2. So any pause greater than 5 seconds would be replaced with a 2 second pause.  The option would allow enable/disable and (if enabled) adjust both times.

**Comment by AESilky** [comment]
Looking at putting the `sleep` in the correct place first (so it comes before the next time).

**Comment by AESilky** [comment]
Sorry if this is a repeat. Seems I lost things in a few open browser windows...

Here is a MKOB and an audio recording of Wire-11 today (20200715):
Audio: https://1drv.ms/u/s!AqnBbsz_ABrPhZU0BndZCzqX6zWsvQ
MKOB: 
[MKOB-Wire11-20200715.zip](https://github.com/MorseKOB/PyKOB/files/5079366/MKOB-Wire11-20200715.zip)

**Comment by AESilky** [comment]
The recording and playback seem to be working, but MKOB is hard-coded to record the session. This issue is still waiting on adding the option to MKOB to record (or not) and then to be able to play back a recording.

Once the record option is added, I might close this and create a separate issue for playing back in MKOB.

**Comment by AESilky** [comment]
Looking at playback timing with the help of @leskerr wiki https://github.com/MorseKOB/PyKOB/wiki/Morse-playback I pulled the 'THE' code sequence from the recording that Les made of 'Now is the time...'. I wanted to compare the timestamp deltas to the playback timing and that discussed in the wiki.

The code sequence for ("STHET") from the recording, as sent by hand, is:
```
'S' {"ts": 1599340646521, "w": 10, "s": "AC - Les - Seattle, WA (Windows) [R]", "o": 1, "c": [-212, 35, -35, 49, -41, 39]}
'T' {"ts": 1599340647044, "w": 10, "s": "AC - Les - Seattle, WA (Windows) [R]", "o": 1, "c": [-350, 172]}
'H' {"ts": 1599340647512, "w": 10, "s": "AC - Les - Seattle, WA (Windows) [R]", "o": 1, "c": [-168, 49, -40, 46, -33, 47, -41, 41]}
'E' {"ts": 1599340647767, "w": 10, "s": "AC - Les - Seattle, WA (Windows) [R]", "o": 1, "c": [-203, 50]}
'T' {"ts": 1599340648295, "w": 10, "s": "AC - Les - Seattle, WA (Windows) [R]", "o": 1, "c": [-331, 197]}
```

Unlike when receiving in real-time from the internet, the code sequences of a recording are all immediately available. The wiki describes the buffering needed before sounding the 'T' because of the time needed to receive the code sequence for the 'H' so that it is ready to sound when the sounding of the 'T' is finished. In a recording the 'T' and the 'H' (and all of the content) are available on demand. Therefore, no buffering is needed to play back the recording.

Comparing the timestamp deltas to the calculated time (sum of the mark/space times), to the measured/actual time to sound the code sequences:
| Char | TS Delta | C Sounder Time [1]| A Sounder Time [2] |
| :-: | -: | -: | -: |
| T | - | 522 | 519 |
| H | 468 | 465 | 462 |
| E | 255 | 253 | 252 |

This illustrates that the timestamp deltas are slightly longer than the calculated and actual playback times. 
Therefore, for code sent without unusual pauses, the code sequences can simply be fed to the KOB player and the sounding should come out correctly.

The place where the timestamp deltas need to come into play is when the user pauses to think about what to send next. This will typically be between words, and more likely between sentences. This shouldn't occur for messages being transcribed, but can happen with interactive conversations. For example, if the operator sends; "WHAT I DID THIS WEEKEND" and then pauses to think about it before sending; "MOWED THE LAWN AND WENT TO THE PARK". To replay the content (for example, as would go onto a telegram) no additional pauses are needed. The code sequence can simply be sent to the KOB sender. However, to replicate the experience of listening to it live, the delay between the two sentences needs to be replicated in the playback.

The timestamps in the recording can be used to calculate the needed 'sleep time' needed to provide the pauses in order to replicate the real-time nature of the transmission. The key is to identify when to introduce the pauses. In addition, an option to reduce very long pauses to a reasonable length is helpful.


T 
```

**Comment by leskerr** [comment]
Here's the smoking gun:

```
C:\Users\Les\Documents\Play>py play.py one-two-three.json
2020-09-07 22:26:10     INFO:   Interface for key and sounder on serial port 'com4' not available. Key and sounder will not be used.
 {"ts": 1599341566538, "w": 10, "s": "AC - Les - Seattle, WA (Windows) [R]", "o": 1, "c": [-32767, 2]}
Sleep:  1.235
 {"ts": 1599341567773, "w": 10, "s": "AC - Les - Seattle, WA (Windows) [R]", "o": 1, "c": [-382, 50, -150, 50]}
 {"ts": 1599341568214, "w": 10, "s": "AC - Les - Seattle, WA (Windows) [R]", "o": 1, "c": [-191, 150, -50, 50]}
 {"ts": 1599341568456, "w": 10, "s": "AC - Les - Seattle, WA (Windows) [R]", "o": 1, "c": [-191, 50]}
 {"ts": 1599341568480, "w": 10, "s": "AC - Les - Seattle, WA (Windows) [R]", "o": 1, "c": []}
 {"ts": 1599341568989, "w": 10, "s": "AC - Les - Seattle, WA (Windows) [R]", "o": 1, "c": [-382, 150]}
 {"ts": 1599341569628, "w": 10, "s": "AC - Les - Seattle, WA (Windows) [R]", "o": 1, "c": [-191, 50, -50, 150, -50, 150]}
 {"ts": 1599341570073, "w": 10, "s": "AC - Les - Seattle, WA (Windows) [R]", "o": 1, "c": [-191, 50, -150, 50]}
Sleep:  1.14
 {"ts": 1599341571653, "w": 10, "s": "AC - Les - Seattle, WA (Windows) [R]", "o": 1, "c": [-191, 150]}
Sleep:  0.388
 {"ts": 1599341572193, "w": 10, "s": "AC - Les - Seattle, WA (Windows) [R]", "o": 1, "c": [-191, 50, -50, 50, -50, 50, -50, 50]}
 {"ts": 1599341572739, "w": 10, "s": "AC - Les - Seattle, WA (Windows) [R]", "o": 1, "c": [-191, 50, -150, 50, -50, 50]}
 {"ts": 1599341572974, "w": 10, "s": "AC - Les - Seattle, WA (Windows) [R]", "o": 1, "c": [-191, 50]}
 {"ts": 1599341573215, "w": 10, "s": "AC - Les - Seattle, WA (Windows) [R]", "o": 1, "c": [-191, 50]}
Sleep:  0.832
 {"ts": 1599341574287, "w": 10, "s": "AC - Les - Seattle, WA (Windows) [R]", "o": 1, "c": [-32767, 1]}
```

The `Sleep: 0.388` shouldn't be there. That's what's causing the extra delay between the T and the H in THREE. The question is, why is it there?

**Comment by AESilky** [comment]
Thanks so much for your superior ear and tracking that down! I will concentrate on that. Finding and fixing that will probably solve some other subtle issues.

**Comment by leskerr** [comment]
I'm working on it now. Not sure what the effect is if I make local changes to your branch. Don't want to step on your toes!

**Comment by AESilky** [comment]
If you get it working, you can just push to GH. I can pick up the changes and test here as well. No problem at all.

**Comment by AESilky** [comment]
I think the recording format is solid. We can continue to 'tweek' the playback.
I'm going to close this and open a new issue to add record/playback to MKOB.

**Comment by leskerr** [comment]
Sounds good to me!

**Comment by AESilky** [comment]
#137 has been created to add 'Record/Play' into the MKOB GUI.

Closing this, as the `recorder` seems to be solid.
